### ISSUE TITLE
Optional entity unmarshalling not working with spray-json

### ISSUE THREAD

**Issue by johanandren** [akka/akka-http#3027]
With an unmarshaller defined for `MyType` the following server route:

```
put(
  entity(as[Option[Order]]) { maybe =>
    complete("Ok")
  }
)
```

when hit with an empty body, responds with: 

```
> PUT / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Type: application/json
> Content-Length: 0
>
< HTTP/1.1 400 Bad Request
< Server: akka-http/10.1.8
< Date: Thu, 26 Mar 2020 14:24:53 GMT
< Content-Type: text/plain; charset=UTF-8
< Content-Length: 122
* HTTP error before end of send, stop sending
<
The request content was malformed:
Unexpected end-of-input at input index 0 (line 1, position 1), expected JSON Value:

^
```

I didn't try with any other unmarshaller. The `targetOptionUnmarshaller` relies on the unmarshalling to fail with `NoContentException` so I'm guessing that is not what spray-json is throwing in this case.

**Comment by jrudolph** [comment]
Yeah, there's also a potential ambiguity that spray-json could itself provide a marshaller for `Option[T]` that would work differently still.

We can `NoContentException` to the spray-json-marshaller but it's somewhat questionable how useful that feature is in general.

I guess a more general solution that would work on any kind of failure could be this:

```scala
entity(as[Order].map(Some(_)).recover(_ => _ => { case x => None }))
```

**Comment by johanandren** [comment]
Any kind of failure is not the same thing as empty body though, it would eat a parse exception as well?

**Comment by jrudolph** [comment]
Yes, it would but the `targetOptionUnmarshaller` is not really specified so we could theoretically think about different interpretations :)

**Comment by johanandren** [comment]
Could it recover taking content length == 0 into account perhaps?
